Use a tiny base image for Docker builds

pugnascotia opened the pull request:
Closes #51670, closes #50838.

Introduce a tiny base image for Docker builds. It aims to create a basic filesystem with as little as possible, which is mostly `glibc`, `busybox` and `bash`. A statically-built `curl` is also provided.

We still use CentOS 8 as a base. All the fun stuff happens in the `Dockerfile`.

**Note:** The comment history below covers a time period when there was an intermediate base image, and a time when the new parts of the build existed in scripts. The current form of the build is a result of discussions with Docker about how official images should be written.

pugnascotia committed:
Dump container logs on shell command failure

The Docker tests framework captures container logs when Elasticsearch
fails to start. However, it doesn't do this if a later shell command
fails. Amend the DockerShell wrapper to capture the container logs if
a command fails when it should succeed.

pugnascotia committed:
Throw original exception docker logs capture

When DockerShell.run() catches a ShellException, dump out the logs and
then rethrow the exception, rather than wrapping the original exception.

pugnascotia committed:
First attempt at building and using a tiny base image

pugnascotia committed:
Build the baseimage in Docker

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

mark-vieira commented:
It'd be nice for changes like this to include a summary of the impact here. Things like:

- Total image size before/after
- Compressed image size before/after
- Total build time (including time to build base image) before/after

We also have a number of items/questions that still need addressing if we choose to peruse this further:

- How do we manage this base image? Do we publish it separately from the ES image or do we just include all the stuff to build it in the Docker context we publish [here](https://github.com/elastic/dockerfiles/tree/7.6/elasticsearch). Would doing the latter be kosher wrt to Dockerhub policies?
- How do we integrate this into the build now that we have to first build a base image. How does this impact build times given that building Docker images is already quite expensive and now we are adding another step.
- Does this have any consequences for Cloud which consumes these images as well?
- Do we expect maintaining what is essentially a forked version of the CentOS image build to be costly going forward?

All and all there are a lot of added complexity here, and that's still w/o all the build integration stuff completely fleshed out. I think we need to consider to end benefit here, which is why I ask for the overall numbers to provide some context in making that decision.

pugnascotia commented:
One thing to bear in mind here is that a reduction in size isn't strictly the goal. Rather, we want to reduce the amount of unnecessary stuff, which can e.g. trigger vulnerability warnings in security scans.

Anyway, the uncompessed image size, as reported by `docker images`, is 792MB on master, and 517MB with these changes, so 275MB smaller.

The compressed size is 383MB before and 290MB after, so 93MB smaller.

On my laptop, with **all** Docker images purged, the time to build the base image is 3m 16s (which included downloading the `centos:7` image). With layer caching it is significantly faster on subsequent builds. That said, since there's nothing secret in the base image, and it's unlikely to change frequently, I had anticipated publishing it to Docker Hub as usual, perhaps by Infra if they were willing to own.

I haven't engaged with Cloud yet as I wanted to validate the approach thus far. They would definitely need to make changes though, since the base image only provides user and group management utilities from BusyBox, which are usable but less full-featured than other utils. The changes to the `Dockerfile` here reflect that. I don't anticipate any changes being difficult though. I already included `pigz` on Cloud's behalf. We'd actually have to engage with them more tightly using this base image, since it doesn't provide a package manager.

In terms on maintenance, I see this similarly to [bitnami/minidev](https://github.com/bitnami/minideb), another stripped-down base distro that Bitnami uses for their Elasticsearch image. We might also choose to use it as a basis for e.g. Logstash at some point down the line.

BTW I haven't forgotten about `docker-slim` - there are some changes in the works that might make it usable.

mark-vieira commented:
> One thing to bear in mind here is that a reduction in size isn't strictly the goal. Rather, we want to reduce the amount of unnecessary stuff, which can e.g. trigger vulnerability warnings in security scans.

Thanks, Rory. That's important context and I would have known that has I actually read #51670 😉 

> That said, since there's nothing secret in the base image, and it's unlikely to change frequently, I had anticipated publishing it to Docker Hub as usual, perhaps by Infra if they were willing to own.

👍 That would save us a lot of redundant effort and headache in our build if we could go this route. We'd need to figure out a lot of details here but IMO that would my preference I think so we could rip out all that additional complexity from our build and basically the only thing that would change in the ES repo is the base image used in our Dockerfile.

pugnascotia commented:
Still looking for feedback on this PR.

joshbressers commented:
Can we still run "yum update" inside the container after these changes?

pugnascotia commented:
No `yum` or `rpm` I'm afraid. It's true that it's handy when you're fooling around in an image to be able to do that, but it adds a lot of MBs to the image, as well as bundling an entire development environment (Python). The main goal is being more careful about what's in the final image, and we'd prefer to avoid shipping unnecessary libraries and binaries.

jasontedor commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

jasontedor committed:
Merge remote-tracking branch 'elastic/master' into tiny-baseimage

jasontedor commented:
@pugnascotia I like the direction here, but I'll withhold a thorough review until after you take this out of draft stage. Let me know it's ready for a thorough review.

pugnascotia committed:
Add Dockerfile for statically building curl

pugnascotia committed:
Merge curl building into tiny image build

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Tweaks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Integrate building a base image into the Docker build

pugnascotia committed:
Allow a platform to be specified

pugnascotia committed:
More improvements

pugnascotia committed:
Build a static curl binary during the Docker build

pugnascotia committed:
Trying to get cross-arch Docker builds to work

pugnascotia committed:
Don't provide curl, and suggest wget instead

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Revert something

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Tweaks

pugnascotia committed:
Add missing license and javadoc

pugnascotia commented:
This PR is now ready for review.

Since official Docker Hub images need to [extend `scratch` or another offical image](https://github.com/docker-library/official-images#repeatability), it isn't practical to build a separate base image and pull it from a registry somewhere. Instead, the base "image" is now just a tar archive of the base filesytem that gets added during the build. This does at least simplify things in some respects. Rebuilding the Docker image, once the base image is created, is now quite a 

I was able to create a Docker image on an EC2 ARM machine and run the Docker packaging tests, but if anyone else feels like having a go, that'd be great. One downside of `FROM scratch` is that cross-building for ARM on e.g. Docker for Mac is that although you can specify `FROM --platform linux/arm64`, I ended up with an `armv7l` image, not `aarch64` image. I don't see this as a problem right now because we intend to build ARM images on an ARM machine.

From clean, building the base image on my laptop takes 1m 11s, and building the docker image takes 33s. On master, building the docker image from clean takes 1m 45s - entertainingly, almost the same time as a base + ES image build, except a rebuild in the base layer case is faster.

You'll notice that the image now includes a `curl` script, which just tells the user to use `wget`. The Core/Infra discussed whether to continue bunding curl, and decided not to. We should, however, document how users can extend the shrunk image, since it's less straightfoward now. I'm happy to add such steps to this PR, though I haven't written them yet.

@mark-vieira I'd particularly appreciate feedback on making the build changes cleaner.

pugnascotia committed:
Use wget instead of curl for docker-compose heathchecks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

mark-vieira reviewed (approved):
Left a few comments but otherwise LGTM.

pugnascotia committed:
Address review feedback

pugnascotia committed:
Compress base filesystem tar with gzip

pugnascotia commented on distribution/docker/build.gradle:
```diff
@@ -174,10 +189,32 @@ task integTest(type: Test) {
+  def platform = "aarch64".equals(architecture) ? 'linux/arm64' : 'linux/amd64'
+  def image = "aarch64".equals(architecture) ? 'arm64v8/centos:7' : 'centos:7'
+
+  final Task buildDockerBaseImageTask = task(taskName("build", architecture, oss, "DockerBaseImage"), type: LoggedExec) {
```
`distribution:docker:buildDockerBaseImage` gives me a 17MB, but then I realised I can gzip it down to 7.3MB and everything still works.

mark-vieira replied:
No need to gzip anything. Gradle gzips task outputs prior to sending to the cache.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia commented on distribution/docker/build.gradle:
```diff
@@ -174,10 +189,32 @@ task integTest(type: Test) {
+  def platform = "aarch64".equals(architecture) ? 'linux/arm64' : 'linux/amd64'
+  def image = "aarch64".equals(architecture) ? 'arm64v8/centos:7' : 'centos:7'
+
+  final Task buildDockerBaseImageTask = task(taskName("build", architecture, oss, "DockerBaseImage"), type: LoggedExec) {
```
I've added tracking for these.

pugnascotia commented on distribution/docker/build.gradle:
```diff
@@ -174,10 +189,32 @@ task integTest(type: Test) {
+  def image = "aarch64".equals(architecture) ? 'arm64v8/centos:7' : 'centos:7'
+
+  final Task buildDockerBaseImageTask = task(taskName("build", architecture, oss, "DockerBaseImage"), type: LoggedExec) {
+    inputs.file(installer)
```
Thanks for the pointer, I've added that.

pugnascotia commented on distribution/docker/build.gradle:
```diff
@@ -174,10 +189,32 @@ task integTest(type: Test) {
+    outputs.cacheIf({ true })
+    executable 'docker'
+    args 'run', '--rm', '-t',
+      '-v', "${projectDir}/build:/build",
```
Changed to `${buildDir}`, I don't know why I didn't do that in the first place 🤷‍♂️

pugnascotia commented on distribution/docker/build.gradle:
```diff
@@ -174,10 +189,32 @@ task integTest(type: Test) {
+void addBuildDockerBaseImage(final String architecture, final boolean oss) {
+  def installer = "${projectDir}/src/docker/bin/install-baseimage.sh"
+  def outputFile = baseImageFilename(architecture)
+  def outputPath = "${projectDir}/build/${outputFile}"
```
Changed.

pugnascotia commented on distribution/docker/src/docker/bin/curl:
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
```
The script won't execute without a shebang, and I don't see how it can throw off anyone else's scripting unless they're doing something daft like `file /usr/bin/curl` (and `file` isn't present in the final image anyway).

mark-vieira replied:
If we know `wget` has funny behavior do we still want to go this route of pointing them at that tool?

pugnascotia replied:
I oscillated about this, but in the end I removed it. If people expect to find `curl`, I'd rather it be missing and they complained. Then maybe we have chance to learn more about what they're trying to do.

pugnascotia committed:
Fix typo that stopped :qa:remote-clusters from logging

pugnascotia committed:
Fix docker-compose healthchecks

pugnascotia committed:
docker-compose fixes

pugnascotia committed:
Remove curl wrapper around wget

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Set GCOS in /etc/password for Docker

pugnascotia committed:
Add a note about extending the shrunk Docker image

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

pugnascotia committed:
Reintroduce a static curl build, only much faster

pugnascotia committed:
Reuse the docker base filesystem for OSS

pugnascotia commented:
I've reintroduced a static `curl` build into this PR, because [elastic/helm-charts](https://github.com/elastic/helm-charts/) and [elastic/cloud-on-k8s](https://github.com/elastic/cloud-on-k8s/) use it for their readiness probes. The new method of building curl is dramatically faster and so far more feasible to add to the build.

I've also added a note to the docs about extending the image.

@rjernst @jasontedor @dliappis I'd also appreciate your input on this PR.

mark-vieira commented:
👍 to continuing to include curl to keep, in particular, existing ECK users to upgrade w/o having to upgrade the operator.

pugnascotia committed:
Docs and permissions fixes

pugnascotia committed:
Upgrade tini and validate its checksum

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

dliappis commented on distribution/docker/src/docker/bin/install-baseimage.sh:
```diff
@@ -0,0 +1,149 @@
+
+# Use busybox instead of installing more RPMs, which can pull in all kinds of
+# stuff we don't want. There's no RPM for busybox available for CentOS.
+BUSYBOX_URL="https://busybox.net/downloads/binaries/1.31.0-i686-uclibc/busybox"
```
At some point there will be bugs with these utilities (busybox, or tini) as well.

I think it's hard to keep track of the versions as they are hard coded here.

What if specify such versions as readonly vars at the top of this shell script so that it's easier to track the versions and do future updates?

example after
```
#!/usr/bin/env bash
...
```

then

```
declare -r BUSYBOX_VERSION="1.31.0"
declare -r TINI_VERSION="0.19.0"
```

dliappis commented on distribution/docker/src/docker/bin/install-baseimage.sh:
```diff
@@ -0,0 +1,149 @@
+# Curl needs files under here. More importantly, we change Elasticsearch's
+# bundled JDK to use /etc/pki/ca-trust/extracted/java/cacerts instead of
+# the bundled cacerts.
+tar cf - /etc/pki | (cd "$target" && tar xf -)
```
Is there a particular reason for the use of `tar` here for copying?

I was thinking something like `mkdir -p "$target"/etc && cp -a /etc/pki "$target"/etc/` is simpler to reason about in the future, or am I missing something?

dliappis reviewed (commented):
Thank you for this massive work!

I left some comments.

pugnascotia commented on distribution/docker/src/docker/bin/install-baseimage.sh:
```diff
@@ -0,0 +1,149 @@
+  --skip-broken \
+  install \
+    java-latest-openjdk-headless \
+    bash nc zip upzip pigz
```
I think `--skip-broken` stops the script from failing. It should, of course, be `unzip`.

pugnascotia committed:
Address review feedback

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
WIP - trying for a cross-arch Docker build

williamrandolph commented:
I am trying to run `./gradlew clean distribution:docker:assemble` on this branch. I'm getting some errors at the `buildDockerImage` and `buildOssDockerImage` stages.

```
| Step 14/25 : RUN addgroup -g 1000 elasticsearch &&     adduser -D -u 1000 -G elasticsearch -g elasticsearch -h /usr/share/elasticsearch elasticsearch &&     addgroup elasticsearch root &&     chmod 0775 /usr/share/elasticsearch &&     chgrp 0 /usr/share/elasticsearch
|  ---> Running in 925713f5df5f
| adduser: /usr/share/elasticsearch: Value too large for defined data type
| Removing intermediate container 925713f5df5f
```

It looks like the issue that's discussed here: https://github.com/GoogleContainerTools/distroless/issues/225

But then I remembered that I had enabled Kubernetes on my Mac's Docker Desktop. I turned that off, restarted Docker, and after that I was able to build and run the small image.

I'm not sure if there's anything to do about this, but if you can understand what happened here maybe there's a way to guard against it. Let me know if I can do more to investigate.

williamrandolph commented on qa/os/src/test/java/org/elasticsearch/packaging/util/ProcessInfo.java:
```diff
@@ -0,0 +1,79 @@
+import static org.hamcrest.Matchers.hasSize;
+
+/**
+ * Encapsulates the fetching of information about a running process.
```
I believe this javadoc should mention that the code here is targeting Linux, and perhaps mention the specific distro or environment it's intended for. The commands here wouldn't work on a Mac (no `/proc` directory, for one thing), and certainly wouldn't work on Windows, and someone new to the code might think this is a general class for retrieving process info.

pugnascotia replied:
Good point, I'll update the Javadoc.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Perform all the build steps via the Dockerfile

This means that Gradle can't cache any of the intermediate steps, but
Docker still can. This change makes it easier for the build to be
cross-architecture.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Explicitly specify packages to install instead of hacky exclusion fiasco

pugnascotia committed:
Tweak comment

pugnascotia committed:
Add more into to ProcessInfo Javadoc

pugnascotia commented:
@williamrandolph thanks for pointing out the Busybox issue. The distroless thread you posted had people reporting that they only saw this problem with the uclibc build of Busybox, so I'll change the build to use the musl build for both architectures.

pugnascotia committed:
Use musl busybox for both archs

pugnascotia committed:
Update docs

pugnascotia committed:
Test tweak

pugnascotia committed:
Follow Docker guidelines and put script contents into Dockerfile

pugnascotia committed:
Trim dependencies further

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix-windows

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia commented:
Can I please ask for a re-review on this PR? It's changed a lot since the first iteration due to feedback from Docker.

mark-vieira reviewed (approved):
LGTM 👍

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia commented:
Josh updated me with a comment:

> I know I suggested to you a while back about not being able to run "yum upgrade" as being potentially problematic for some customers. In the last month or so I've gotten drastically more requests for a dockerfile than the ability to run yum upgrade. The tide is turning toward small containers over the ability to update them.

pugnascotia commented:
I've engaged with Cloud Delivery to begin some testing of the revised image. We'll figure out what it takes for Cloud to consume it, and flush out any issues.

mieciu commented on distribution/docker/src/docker/Dockerfile:
```diff
@@ -10,70 +10,174 @@
+RUN yum --installroot=/rootfs --releasever=/ --setopt=tsflags=nodocs \\
+      --setopt=group_package_types=mandatory -y  \\
+      --skip-broken \\
+      install basesystem bash zip zlib
```
How about adding `pigz`? 
It's very useful for compressing large heap dumps. It's not a must-have from Cloud's perspective (we can fall back to `gzip`) but since we will be lacking package manager it'd make a lot of sense to add it here.

jasontedor replied:
@mieciu It's my assumption that heap dumps would be written to a volume that is mounted on the host (otherwise, they vanish when the container dies, which will happen when the process died because of an OOME), and thus could be processed on the host external to the container.

pugnascotia replied:
That's a good point - @mieciu what is Cloud doing with heapdumps exactly - compressing them in the container?

joegallo replied:
We could move the compression out of the ES container and into the allocator container. Then pigz is always available (if we want it to be).

pugnascotia replied:
So I'm happy to do something to smooth the way, though compressing heapdumps does feel like an orchestration concern, so ultimately I think the allocator should handle it in whatever way it deems fit.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia committed:
Tweaks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia commented:
I've updated this PR with `master`, which unfortunately means the `Dockerfile` now combines the UBI work and the tiny image work. It's not pretty, and we could split it so that we have separate files for UBI and non-UBI...but there are steps in common, which would increase the maintenance burden. I'l like to hear everyone's opinion here, particularly @mark-vieira, @rjernst and @jasontedor.

mark-vieira commented:
I don't think it's a big deal. I prefer this to duplication since Docker doesn't have a good solution for "sharing" logic aside from base images which isn't ideal for this use case.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

mieciu reviewed (approved):
**LGTM** in behalf of Cloud.
We're ready to incorporate those changes 👍

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into tiny-baseimage

pugnascotia commented:
@jasontedor Latest stats:

| *Measure* | *CentOS 8* | *Scratch* |
| ------ | ---------- | --------- |
| `docker images` size | 771MB | 558MB |
| Number of files | 27829 | 18797 |
| Number of executables in `/usr/{bin,sbin}` | 582 | 431 |
| `docker save ... \| gzip` size | 379M | 311M |

In terms of reducing attack surface, we would ship about a third fewer files, notably omitting the standard CentOS stuff like systemd and `yum`, as well as `python` and a bunch of libraries.

We should see a greater size reduction on ARM, because the CentOS image ships a whole bunch of device drivers (at least I think that's what they are?), which adds hundreds of megabytes to the image. I actually can't remember when I last tried this build on ARM, however, so I should do that before merging.

pugnascotia commented:
OK, I spun up an ARM images in AWS and checked that the branch builds OK, which it does. The `docker images` size is about the same. I forgot to capture any more stats before I terminated the instance, unfortunately, but I don't expect much variation.

Entertainingly, the ARM version of `centos:8` is now only `255MB`, vs `215MB` for `x86_64`, whereas I'm pretty sure it was much higher for `centos:7`. But still, we're saving space.

So...I guess we're good to go. Eek.

michelkluger commented:
is there a slim/alpine version of elastic search I could use for my CI? instead of pulling 640mb?
